rfcs/text/0021-threat-enrichment.md

```diff
@@ -39,12 +50,130 @@ Stage 2: Add or update all remaining field definitions. The list should now be e
+
+If it is determined that an event matches a given indicator, that event can be enriched with said indicator. Presently, Indicator Match Detection rules will perform that enrichment automatically, but ad hoc/manual enrichment is an expected feature that analysts will leverage in the future.
+
+#### Example document
+
+```json5
+{
+  "process": {
+    "name": "svchost.exe",
+    "pid": 1644,
+    "entity_id": "MDgyOWFiYTYtMzRkYi1kZTM2LTFkNDItMzBlYWM3NDVlOTgwLTE2NDQtMTMyNDk3MTA2OTcuNDc1OTExNTAw",
+    "executable": "C:\\Windows\\System32\\svchost.exe"
+  },
+  "message": "Endpoint file event",
+  "@timestamp": "2020-11-17T19:07:46.0956672Z",
+  "file": {
+    "path": "C:\\Windows\\Prefetch\\SVCHOST.EXE-AE7DB802.pf",
+    "extension": "pf",
+    "name": "SVCHOST.EXE-AE7DB802.pf",
+    "hash": {
+      "sha256": "0c415dd718e3b3728707d579cf8214f54c2942e964975a5f925e0b82fea644b4"
+    }
+  },
+  "threat": {
+    "enrichments": [
+      {
+        // Each enrichment is added as a nested object under `threat.enrichments.*`
+        // Copy all the object indicators under `indicator.*`, providing full context
+        "indicator": {
+          "marking": {
+            "tlp": "WHITE"
+          },
+          "first_seen": "2020-10-01",
+          "last_seen": "2020-11-01",
+          "sightings": 4,
+          "type": ["sha256", "md5", "file_name", "file_size"],
+          "description": "file last associated with delivering Angler EK"
+        },
+        // event and file fields are copied from the indicator doc, if relevant
+        "event": {
+          "provider": "Abuse.ch",
+          "dataset": "threatintel.abusemalware",
+          "module": "threatintel"
+        },
+        "file": {
+          "hash": {
+            "sha256": "0c415dd718e3b3728707d579cf8214f54c2942e964975a5f925e0b82fea644b4",
+            "md5": "1eee2bf3f56d8abed72da2bc523e7431"
+          },
+          "size": 656896,
+          "name": "invoice.doc"
+        },
+        /* `matched` will provide context about which of the indicators above matched on this
+              particular enrichment. If multiple matches for this indicator object, this could
+              be a list */
+        "matched": {
+          "atomic": "0c415dd718e3b3728707d579cf8214f54c2942e964975a5f925e0b82fea644b4",
+          "field": "file.hash.sha256",
+          "id": "abc123f03",
+          "index": "threat-indicators-index-000001",
+          "type": "indicator_match_rule"
+        }
+      }
+    ]
+  },
+  // Tag the enriched document to indicate the threat enrichment matched
+  // TODO I think this is now redundant with matched.type, but maybe still useful?
+  "tags": ["threat-match"],
+  // This should already exist from the original ingest pipeline of the document
+  "related": {
+    "hash": ["0c415dd718e3b3728707d579cf8214f54c2942e964975a5f925e0b82fea644b4"]
+  }
+}
+```
+
+### Proposed enrichment pipeline mechanics pseudocode
```

Comment:
@dcode @peasead I know this section is out of date, but I had some trouble grokking and could use some help on this one.

Reply:
@dcode, @rylnd and I chatted about this, but may still need some specifics if you could provide those.

Reply:
@rylnd did you get the information you needed on this?

Reply:
@peasead I think I finally grokked and updated that section appropriately. If you and @dcode can verify that what I wrote still makes sense, lemme know 👍